crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl

```diff
@@ -48,61 +73,60 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
     if !lod_is_ok || parent_lod_is_ok { return; }
+#endif
 
-    // In the first pass, operate only on the clusters visible last frame. In the second pass, operate on all clusters.
-#ifdef MESHLET_SECOND_CULLING_PASS
-    var meshlet_visible = true;
+    // Project the culling bounding sphere to view-space
+#ifdef MESHLET_FIRST_CULLING_PASS
+    let culling_bounding_sphere_center_view_space = (previous_view.inverse_view * vec4(culling_bounding_sphere_center.xyz, 1.0)).xyz;
+    let previous_model = affine3_to_square(instance_uniform.previous_model);
+    let previous_model_scale = max(length(previous_model[0]), max(length(previous_model[1]), length(previous_model[2])));
+    culling_bounding_sphere_center = previous_model * vec4(bounding_spheres.self_culling.center, 1.0);
+    culling_bounding_sphere_radius = previous_model_scale * bounding_spheres.self_culling.radius;
 #else
-    var meshlet_visible = get_meshlet_previous_occlusion(cluster_id.x);
-    if !meshlet_visible { return; }
+    let culling_bounding_sphere_center_view_space = (view.inverse_view * vec4(culling_bounding_sphere_center.xyz, 1.0)).xyz;
 #endif
 
-    // Calculate world-space culling bounding sphere for the cluster
-    let culling_bounding_sphere_center = model * vec4(bounding_spheres.self_culling.center, 1.0);
-    let culling_bounding_sphere_radius = model_scale * bounding_spheres.self_culling.radius;
-
-    // Frustum culling
-    // TODO: Faster method from https://vkguide.dev/docs/gpudriven/compute_culling/#frustum-culling-function
-    for (var i = 0u; i < 6u; i++) {
-        if !meshlet_visible { break; }
-        meshlet_visible &= dot(view.frustum[i], culling_bounding_sphere_center) > -culling_bounding_sphere_radius;
+    let aabb = project_view_space_sphere_to_screen_space_aabb(culling_bounding_sphere_center_view_space, culling_bounding_sphere_radius);
+    let depth_pyramid_size_mip_0 = vec2<f32>(textureDimensions(depth_pyramid, 0)) * 0.5; // Halve the view-space AABB size as the depth pyramid is half the view size
+    let width = (aabb.z - aabb.x) * depth_pyramid_size_mip_0.x;
+    let height = (aabb.w - aabb.y) * depth_pyramid_size_mip_0.y;
+    let depth_level = max(0, i32(ceil(log2(max(width, height))))); // TODO: Naga doesn't like this being a u32
+    let depth_pyramid_size = vec2<f32>(textureDimensions(depth_pyramid, depth_level));
+    let aabb_top_left = vec2<u32>(aabb.xy * depth_pyramid_size);
+
+    let depth_quad_a = textureLoad(depth_pyramid, aabb_top_left, depth_level).x;
+    let depth_quad_b = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 0u), depth_level).x;
+    let depth_quad_c = textureLoad(depth_pyramid, aabb_top_left + vec2(0u, 1u), depth_level).x;
+    let depth_quad_d = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 1u), depth_level).x;
+    let occluder_depth = min(min(depth_quad_a, depth_quad_b), min(depth_quad_c, depth_quad_d));
```

Comment:
I know it's not new code but could/shouldn't this use: https://www.w3.org/TR/WGSL/#texturegather ?

Reply:
textureGather doesn't allow you to specify mip leve, so we can't.